include/wil/enum_iterator.h

```diff
@@ -0,0 +1,146 @@
+
+        template<typename T, typename Arg>
+        struct FunctionTraits<HRESULT(__stdcall T::*)(ULONG, Arg**, ULONG*)>
+            : FunctionTraitsBase<HRESULT, T, ULONG, Arg, ULONG*>
+        {};
+
+#ifdef __cpp_noexcept_function_type
+        template<typename T, typename Arg>
+        struct FunctionTraits<HRESULT(__stdcall T::*)(ULONG, Arg**, ULONG*) noexcept>
+            : FunctionTraitsBase<HRESULT, T, ULONG, Arg, ULONG*>
+        {};
+#endif
+
+        template<typename TEnum>
+        using NextArgType = typename Details::FunctionTraits<decltype(&TEnum::Next)>::template NthArg<1>;
+    }
+
+    template<typename TEnum>
+    /// <summary>
+    /// Iterator class for iterating through types that implement the IEnum* COM pattern.
+    /// </summary>
+    /// <typeparam name="TEnum">The enumerator type, e.g. IEnumIDList, IDiaEnumSymbols.</typeparam>
+    struct enum_iterator
+    {
+        enum_iterator(const enum_iterator& other) = default;
+        enum_iterator(TEnum* pEnum) : m_pEnum(pEnum)
+        {
+            (*this)++;
+        }
+        /// <summary>
+        /// The type of the out parameter in Next(...) that corresponds to the element type being enumerated.
+        /// e.g. PITEMID_CHILD, IDiaSymbol, etc.
+        /// </summary>
+        using TElem = Details::NextArgType<TEnum>;
+        ~enum_iterator() = default;
+
+        auto& operator++(int) { return this->operator++(); }
+        auto& operator++()
+        {
+            TElem* pChild{ nullptr };
+            ULONG celt = 0;
+            auto hr = m_pEnum->Next(1, &pChild, &celt);
+            if ((hr == S_OK) && (celt == 1))
+            {
+                m_current = pChild;
+            }
+            else if (hr == S_FALSE)
+            {
+                m_end = true;
+            }
+            else
+            {
+                THROW_HR(hr);
+            }
+            return *this;
+        }
+
+        TElem* const& operator*() const noexcept { return m_current; }
+        TElem* operator->() const noexcept { return m_current; }
+
+        auto operator+(int v)
+        {
+            auto other = *this;
+            for (int i = 0; i < v; i++)
+            {
+                other++;
+            }
+            return other;
+        }
+
+        bool operator==(const enum_iterator& o) const noexcept
+        {
+            if (m_end ^ o.m_end) return false;
+            if ((m_end == o.m_end) && m_end) return true;
+            return (o.m_current == this->m_current) && (o.m_pEnum == m_pEnum) && (m_end == o.m_end);
+        }
+
+        bool operator!=(const enum_iterator& o) const noexcept { return !(*this == o); }
+
+        static constexpr auto end() noexcept
+        {
+            return enum_iterator{ end_tag{} };
+        }
+    protected:
+        struct end_tag {};
+        enum_iterator(end_tag) : m_end(true) {}
+        enum_iterator() {};
+        wil::com_ptr<TEnum> m_pEnum{ nullptr };
+        TElem* m_current{ nullptr };

```

Comment:
What about collections of scalars?

Reply:
yeah I need to think about it more; esp. other non-COM types that still need to be resource-managed (like ITEMID_CHILD needs to be ComTaskFree'd)